src/bloqade/native/upstream/squin2native.py

```diff
@@ -0,0 +1,128 @@
+from dataclasses import field, dataclass
+
+from kirin import ir, passes, rewrite
+from kirin.dialects import py, func
+from kirin.rewrite.abc import RewriteRule, RewriteResult
+from kirin.passes.callgraph import CallGraphPass, ReplaceMethods
+from kirin.analysis.callgraph import CallGraph
+
+from bloqade.native import kernel, broadcast
+from bloqade.squin.clifford import stmts, dialect as clifford_dialect
+
+
+class GateRule(RewriteRule):
+    SQUIN_MAPPING: dict[type[ir.Statement], tuple[ir.Method, ...]] = {
+        stmts.X: (broadcast.x,),
+        stmts.Y: (broadcast.y,),
+        stmts.Z: (broadcast.z,),
+        stmts.H: (broadcast.h,),
+        stmts.S: (broadcast.s, broadcast.s_adj),
+        stmts.T: (broadcast.t, broadcast.t_adj),
+        stmts.SqrtX: (broadcast.sqrt_x, broadcast.sqrt_x_adj),
+        stmts.SqrtY: (broadcast.sqrt_y, broadcast.sqrt_y_adj),
+        stmts.Rx: (broadcast.rx,),
+        stmts.Ry: (broadcast.ry,),
+        stmts.Rz: (broadcast.rz,),
+        stmts.CX: (broadcast.cx,),
+        stmts.CY: (broadcast.cy,),
+        stmts.CZ: (broadcast.cz,),
+    }
+
+    def rewrite_Statement(self, node: ir.Statement) -> RewriteResult:
+        if (native_methods := self.SQUIN_MAPPING.get(type(node))) is None:
+            return RewriteResult()
+
+        if isinstance(node, stmts.SingleQubitNonHermitianGate):
+            native_method = native_methods[1] if node.adjoint else native_methods[0]
+        else:
+            native_method = native_methods[0]
+
+        # do not rewrite in invoke because callgraph pass will be looking for invoke statements
+        (callee := py.Constant(native_method)).insert_before(node)
+        node.replace_by(func.Call(callee.result, tuple(node.args), kwargs=()))
```

Comment:
If I understand correctly, you are replacing each statement with an actual `invoke` to a stdlib function. On the one hand, this is quite elegant since it re-uses a lot of code. On the other hand, however, I'm not sure if this leads to the best IR here.

Since we use stdlib everywhere (also in clifford), it's quite likely that the statements you are rewriting are already inside an invoke of a `clifford.stdlib` function. So you will end up nesting a lot of invokes.

Maybe we can inline the newly added invokes here? You could collect the hashes of the invokes you insert and inline them in the end with an appropriate `heuristic`.

Reply:
Actually this is more efficient if you are only using the clifford standard libraries I will rewrite those subroutines and not the invoke of the subroutine. the complexity is bounded by the number of standard library functions. not by the number of invokes of those standard library functions.